docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,249 @@
+have a new lease granted on it.
+
+Lease acquisition will perform the following steps in a transaction:
+
+* `SELECT descriptor FROM system.descriptor WHERE id = <descID>`
+* `INSERT INTO system.lease VALUES (<desc.ID>, <desc.version>, <nodeID>, <expiration>)`
+
+Nodes will maintain a map from `<descID, version>` to
+`<TableDescriptor, expiration, localRefCount>`. The local reference
+count will be incremented when a transaction first uses a table and
+decremented when the transaction commits/aborts. When the node
+discovers a new version of the table, either via gossip or by
+acquiring a lease and discovering the version has changed it can
+release the lease on the old version when there are no more local
+references:
+
+* `DELETE FROM system.lease WHERE (DescID, Version, NodeID, Expiration) = (<descID>, <version>, <nodeID>, <expiration>)`
+
+A schema change operation needs to ensure that there is only one
+version of a descriptor in use in the cluster before a mutation of the
+descriptor. The operation will perform the following steps
+transactionally:
+
+* `SELECT descriptor FROM system.descriptor WHERE id = <descID>`
+* `SELECT DISTINCT version FROM system.lease WHERE descID = <descID> AND expiration > current_time()`
+* Ensure that there is only one active version which is equal to the current version.
+* Perform the edits to the descriptor and bump the descriptor version.
+* `UPDATE system.descriptor WHERE id = <descID> SET descriptor = <desc>`
+
+Note that the updating of the table descriptor will cause the table
+version to be gossipped alerting nodes to the new version and causing
+them to release leases on the old version. The expectation is that
+nodes will fairly quickly transition to using the new version and
+release all leases to the old version allowing another step in the
+schema change operation to take place. The schema change operation
+must poll the `system.lease` table waiting for the number of active
+versions of a descriptor to fall to 1 or 0. This polling will be
+performed at a short interval (1s), perhaps with a small amount of
+back-off.
+
+When a node holding leases dies permanently or becomes unresponsive
+(e.g. detached from the network) schema change operations will have to
+wait for any leases that node held to expire. This will result in an
+inability to perform more than one schema modification step to the
+descriptors referred to by those leases. The maximum amount of time
+this condition can exist is the lease duration (5m). Note that
+attempting to release leases granted by a previous incarnation of a
+node is not safe because an in-flight but uncommitted transaction
+using those leases might still exist in the system.
+
+As described above, leases will be retained for the lifetime of a
+transaction. In a multi-statement transaction we need to ensure that
+the same version of each table is used throughout the transaction. To
+do this we will add the descriptor IDs and versions to the transaction
```

Comment:
which transaction structure? I'm going to assume the state the sql server holds about the user transaction (not adding this information to the txn proto).